### ISSUE TITLE
Void semantics are not quite clear

### ISSUE THREAD

**Issue by flbulgarelli** [uqbar-project/wollok#701]
Testing void methods, I found something a little disturbing. 

Lets suppose I have the following code: 

```
object foo {
    method bar() {}
}

program abc {
    console.println(foo.bar()) ;
}
```

Then I notice that this method, as it is `void`, produces no value. And the editor nicely detects that - `This message call produces no value`. Not only that, but also the interpreter rejects this program. :+1:

But then, I try the following: 

```
program abc {
    var otherFoo = foo;
    console.println(otherFoo.bar()) ;
}
```

I see no error in the editor. So i think:

> The editor is not yet clever enough to detect this situation. And it doesn't notice that this will fail. This is nothing strange at all - it is exactly what happens when you try to send a message to an object that does not understand it, and the editor does not have enough information to notice that. But it will fail on runtime. 

But then I execute the code. To my surprise, it works! And prints `void[]`

I am now a bit astonished: the editor was rejecting a program that was valid! I know that a checker can not filter all invalid program, and it can also filter valid programs. The first scenario is quite common and acceptable. But the latter is something that is posible, but undesirable, and you wan't to be away from such scenarios - think about Java and their generics headaches. It introduces extra complexity to the user model. 

As a rule of thumb, when building robust systems, if validation fails, so will runtime system in case you manage skip that validation. Otherwise, it should be a warning at most. 

 But, let's suppose you don't agree with the previous, lacking-of-references, rule. I find another problem which is more profound: the semantics of void are not well defined: the first error message states `This message call produces no value`, which means that a void method returns nothing, and nothing can you get from such method. 

But  then, you actually returned something! And it is a value, you can pass and print. That abstraction is leaky.   

I think that either the first scenario should just produce a warning with a more sincere message `warning: sending this message will return no useful object`, or using void should not be possible, and assigning or sending messages to void should produce errors in runtime. 

I _suspect_ that the latter semantics may be better.

**Comment by asanzo** [comment]
We talked about this during yesterday's Wollok Hackaton. We used a very simmilar example:

First, the things that work properly (using current new project example):
```Wollok
>>> var h = 3
3
>>> h = ex.sayHi()
ERROR: Este mensaje no retorna ningún valor (linea: 1)
>>> h
3
```
 :+1: 

Now, the problematic:

```Wollok
>>> var ex2 = ex
ex[]
>>> var h = ex2.sayHi()
Hello Wollok!
>>> h
>>> 
>>> h == void
true
```

We see 3 things happening here:

1. The already mentioned problem that `void` is being returned as a result of a method that shouldn't return anything.
2. The method "sayHi()" is actually being executed.
3. The word `void` is treated as a valid value when one writes "void".

With @npasserini we thought that a possible way to fix this is to **check in runtime before executing each method** if the method is expected to return something in the context of its usage, but in its implementation it doesn't return. Then, a runtime error should be thrown.

This means that 1. and 2. will be solved, because:

```Wollok
>>> var h = ex2.sayHi()
```

won't print "Hello Wollok!", but throw an error and `h` would never change its value.
Also, Franco's program:

```Wollok
program abc {
    var otherFoo = foo;
    console.println(otherFoo.bar()) ;
}
```

Won't print anything to the console, because the println would never execute as `otherFoo.bar()` will throw an error for being used as a value returning expression when it is not.
